### PR TITLE
FSL randomise is now parallelized.

### PR DESCRIPTION
**Pull request by natacha-beck**

### CODE REVIEW

**Comment by prioux** [comment]
This pull request is made to work with the new subtask submission system of CBRAIN https://github.com/aces/cbrain/pull/355

**Comment by natacha-beck** [comment]
This works was done by @prioux and me. But for simplicity I do the PR.

File: bin/fsl_sub [hunk 1]
```diff
@@ -95,9 +98,26 @@ if [ "X$FSLSUBALREADYRUN" = "Xtrue" ] ; then
...
+#
+# Triggered by having the following
+# environment variables set:
+#   - CBRAIN_WORKDIR
+# Also uses optionally:
+#   - CBRAIN_SHARE_WD_TID
+#############################
+if test -n "$CBRAIN_WORKDIR" ; then
+  METHOD="CBRAIN"
+fi
```
Comment (glatard) [inline bin/fsl_sub]: Breaks the protection that prevents jobs to submit new jobs (FSLSUBALREADYRUN).

File: bin/fsl_sub [hunk 2]
```diff
@@ -488,58 +527,99 @@ EOF2
...
+
+BASH_SCRIPT
+
+          # Create the JSON descriptor
+          cat <<JSON_FILE >"$prefix.json"
 {
 "tool-class": "CbrainTask::FslSub",
-"description": "$*",
+"description": "$JobName",
+"share_wd_tid": "$CBRAIN_SHARE_WD_TID",
```
Comment (glatard) [inline bin/fsl_sub]: Would be cleaner if it was called 'share-wd-tid' (but has to be modified consistently in cluster_task.rb, see comment in https://github.com/aces/cbrain/pull/355).

File: bin/fsl_sub [hunk 3]
```diff
@@ -488,58 +527,99 @@ EOF2
...
+JSON_FILE
 
           # Waits for the task id for up to 300 seconds
-          TASK_FILE="${FILE%%.*}.cbid"
-          x=0
-          while [ "$x" -lt 300 -a ! -e ${TASK_FILE} ]; do
-            x=$((x+1))
-            sleep 1
+          cbrain_id_file="$prefix.cbid"
+          SECONDS=0 # bash auto-increment magical variable magic magic
```
Comment (glatard) [inline bin/fsl_sub]: Nice :+1:

File: bin/fsl_sub [hunk 4]
```diff
@@ -488,58 +527,99 @@ EOF2
...
-                line=`sed -n -e ''${n}'p' $taskfile`
-                if [ $verbose -eq 1 ] ; then
-                    echo Submitting to CBRAIN: $line >&2
-                fi
-                submit_cbrain ${line}
-                n=`expr $n + 1`
-            done
+            subtask_counter=0
+            cb_subtask_ids=""
+            cat "$taskfile" | perl -ne 'print unless /^\s*$|^\s*#/' > "$taskfile.cleaned"
```
Comment (glatard) [inline bin/fsl_sub]: Do we need to add a dependency on Perl for doing that?

File: cbrain_task_descriptors/fsl_sub.json [hunk 5]
```diff
@@ -1,16 +1,16 @@
 {
   "name": "fsl_sub",
-  "tool-version": "1.1",
+  "tool-version": "2.0",
   "description": "A wrapper for FSL sub. Assumes that task will run on a file system shared with the submission process, that input files are already in place, and that output files will be handled by another process.",
-  "command-line": "fsl_sub [FSL_SUB_OPTIONS]",
+  "command-line": "/bin/bash .new-task-[SUBTASK_ID].sh",
```
Comment (glatard) [inline cbrain_task_descriptors/fsl_sub.json]: Works in our case but this completely bypass fsl_sub for tasks executed with METHOD=NONE.

File: bin/fsl_sub (hunk 2 above)
Comment (glatard) [inline bin/fsl_sub]: This is modified in cluster_task so now it *has to* be modified here.

**Pull request merged**